Improve byte and boolean stream readers

yingsu00 opened the pull request:
BenchmarkSelectiveStreamReaders shows up to 3.8x for byte type, 3.4x for boolean type



```
== NO RELEASE NOTE ==
```

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -160,7 +165,13 @@ public int read(int offset, int[] positions, int positionCount)
 
-        if (outputRequired) {
+        if (useBatchMode()) {
+            // values need to be totalPositionCount,
```
Minor: would be nice to reword and reformat this comment. ByteSelectiveStreamReader has a nice wording for this comment.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -181,6 +192,59 @@ else if (filter == null) {
+                        dataStream.getSetBits(readCount, values);
+
+                        if (nullCount != 0) {
+                            // Note it should be totalPositionCount instead of positionCound
```
typo in positionCound

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -175,6 +184,58 @@ else if (filter == null) {
+                    dataStream.next(values, readCount);
+
+                    if (nullCount != 0) {
+                        // Note it should be totalPositionCount instead of positionCound
```
typo: positionCound

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -456,4 +539,60 @@ public void close()
+    {
+        // JMH benchmark shows that when there is no null or no filter, the batch read mode was always better than the skipping mode.
+        // When there is filter and partial nulls, the batch read mode was better than the skipping mode when the input filter rate is less than 0.4
+        if (presentStream == null || filter == null || (double) (totalPositionCount - positionCount) / totalPositionCount <= 0.4) {
```
We won't hit this branch when totalPositionCount = 0, right?
Can replace the condition with `(double)positionCount/totalPositionCount >= 0.6`

yingsu00 replied:
> We won't hit this branch when totalPositionCount = 0, right?
Right. positionCount and totalPositionCount must be > 0 

> If you want you can replace the condition with `(double)positionCount/totalPositionCount >= 0.6`
@sdruzkin Thanks for the suggestion. I was trying to align with the "filter rate" convention in Presto query plans, and the filter rate in `BenchmarkSelectiveStreamReaders`. It'll be easier to match and reason about if we use the same meaning and number here.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ReaderUtils.java:
```diff
@@ -98,6 +98,24 @@ public static void packByteArrayAndOffsets(byte[] values, int[] offsets, int[] p
+    public static void packBytes(byte[] values, int[] positions, int positionCount)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            values[i] = values[positions[i]];
```
Does this function always moves values from the right side of the array to the left side?

yingsu00 replied:
Yes, positions array is monotonically increasing.

sdruzkin commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -1518,8 +1518,9 @@ public static void writeOrcColumnsPresto(File outputFile, Format format, Compres
         metadata.put("columns", String.join(", ", columnNames));
         metadata.put("columns.types", createSettableStructObjectInspector(types).getTypeName());
 
+        FileOutputStream fileOutputStream = new FileOutputStream(outputFile);
```
Would be nice to use the closer or try-with-resource for fileOutputStream.

sdruzkin reviewed (approved):
Nice job. 

Average improvement w/o nulls is 1.8x (p50 1.7x), w/ nulls 1.2x (p50 1.2x), with a regression to 0.85x for some cases.

yingsu00 commented:
Verifier for this PR and https://github.com/prestodb/presto/pull/15496 has passed https://www.internalfb.com/intern/presto/verifier/results/?test_id=47949
cc @mbasmanova

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -69,6 +70,11 @@
 
     boolean testBoolean(boolean value);
 
+    default boolean testBoolean(byte value)
```
can you please add a javadoc on why this testBoolean method with byte value ? Looking at the surface, I expected this method to be testByte.

mbasmanova replied:
conventions: Don't add default implementation. Just add implementation in all places like other methods.

arunthirupathi reviewed (approved):
Looks good to me.

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:
```diff
@@ -266,7 +268,7 @@ else if (typeSignature.startsWith("varchar")) {
 
             // Use writeOrcColumnsPresto so that orcType and varchar length can be written in file footer
-            writeOrcColumnsPresto(orcFile, ORC_12, NONE, Optional.empty(), Collections.nCopies(channelCount, type), values, new OrcWriterStats());
+            writeOrcColumnsPresto(orcFile, DWRF, NONE, Optional.empty(), Collections.nCopies(channelCount, type), values, new OrcWriterStats());
```
Why is this change? It needs to go in a separate commit if it is needed.

yingsu00 replied:
It was a leftover from testing integer types and it's not related to byte and boolean. I have removed the change.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -456,4 +538,62 @@ public void close()
+                i++;
+            }
+            else {
+                i += filter.getSucceedingPositionsToFail() + 1;
```
why +1 ? is this code path coved by tests?

yingsu00 replied:
the +1 is for the current position. For example, if an array has 8 elements (position 0 to 7) and position 1 fails, then getSucceedingPositionsToFail is 6, and getPrecedingPositionsToFail is 1. To skip to the next array (position 8), `i` need to be advanced with the following 6 elements and itself, therefore the +1. 

Array of integers are tested in existing test but there is no test for array of other types. I added  a new test for array of bytes in `testArrays()` in `TestSelectiveOrcReader`
```
       // non-null arrays of TINYINT
        tester.testRoundTrip(arrayType(TINYINT),
                createList(NUM_ROWS, i -> randomBytes(8, random)),
                ImmutableList.of(
                        ImmutableMap.of(new Subfield("c[2]"), IS_NULL),
                        ImmutableMap.of(new Subfield("c[2]"), nonNegative)));
```

mbasmanova replied:
I get it now. I was confused because I missed that you are using while and not for loop. I think a for loop would be clearer.

```
        for (int i = 0; i < positionCount; i++) {
           int position = positions[i];
            if (filter.testLong(values[position])) {
                outputPositions[positionsIndex++] = position;  // compact positions on the fly
            }
            else {
                i += filter.getSucceedingPositionsToFail();
                positionsIndex -= filter.getPrecedingPositionsToFail();
            }
        }
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -175,6 +182,59 @@ else if (filter == null) {
+        if (useBatchMode(positionCount, totalPositionCount)) {
+            int readCount = 0;
+
+            final int filteredPositionCount;
```
conventions: don't use final

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -175,6 +182,59 @@ else if (filter == null) {
     private int readWithFilter(int[] positions, int positionCount)
             throws IOException
     {
+        int totalPositionCount = positions[positionCount - 1] + 1;
```
I don't see new tests to cover new code paths. Are they all covered by the existing tests?

yingsu00 replied:
Yes. They are already covered in `testByteValues()`, `testByteValuesRepeat()`, `testByteValuesPatchedBase()` and `testBooleanSequence()`.  Skipping will be covered by new tests mentioned above in `testArrays()` for boolean and byte types.

mbasmanova reviewed (commented):
@yingsu00 The commit message says "and also changed the file format from ORC
to DWRF". This is no longer the case. Please, update.

yingsu00 commented:
> @yingsu00 The commit message says "and also changed the file format from ORC
> 
> to DWRF". This is no longer the case. Please, update.

Thanks for the catch! Just updated.

yingsu00 commented:
@mbasmanova comments addressed and tests added, will you be able to take another look? The verifier has passed. Thank you very much!

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -175,6 +182,59 @@ else if (filter == null) {
+                    dataStream.next(values, readCount);
+
+                    if (nullCount != 0) {
+                        // Note it should be totalPositionCount instead of positionCound
```
@yingsu00 I still see "positionCound", not "positionCount".

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java:
```diff
@@ -456,4 +538,62 @@ public void close()
+
+    private int evaluateFilterWithNulls(int[] positions, int positionCount, boolean testNull)
+    {
+
```
@yingsu00 I still see 2 extra empty lines here.
